src/ORM/Connect/DuplicateEntryException.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace SilverStripe\ORM\Connect;
+
+/**
+ * Exception for errors related to duplicate entries (e.g. violating a unique index)
+ */
+class DuplicateEntryException extends DatabaseException
+{
+    private ?string $keyName = null;
+
+    private ?string $duplicatedValue = null;
+
+    /**
+     * Constructs the database exception
+     *
+     * @param string $message The Exception message to throw.
+     * @param string|null $keyName The name of the key which the error is for (e.g. index name)
+     * @param string|null $duplicatedValue The value which was duplicated (e.g. combined value of multiple columns in index)
+     * @param string|null $sql The SQL executed for this query
+     * @param array $parameters The parameters given for this query, if any
+     */
+    public function __construct(
+        string $message = '',
+        ?string $keyName = null,
+        ?string $duplicatedValue = null,
+        ?string $sql = null,
+        array $parameters = []
+    ) {
```

Comment:
Doesn't use `$code` or `$previous` because those were unused in core usages of the parent class anyway.